Simplifying elm-mode

theothornhill opened the pull request:
Hello again!

As discussed in #175, let's start discussing how to merge/simplify this. 

With these commits, things shouldn't really be that controversial.

What I've done:
- Remove stuff that's obviously not needed since `elm-language-server`
- Add my own indentation logic - not enabled yet. 
- Don't remove `elm-interactive` just yet, though most of it can go

To enable my indentation logic, I've added this snippet to my own config:

```elisp
(use-package elm-mode
  :ensure nil
  :bind (:map elm-mode-map
              ("C-c TAB" . #'elm-indent-line)
              ("<tab>" . #'elm-indent-forward)
              ("<backtab>" . #'elm-indent-backward)))

(remove-hook 'elm-mode-hook 'elm-indent-mode)
```

This indentation system doesn't try to be smart at all. There are some defaults for the easiest cases that are simple to get right, other than that, just use `tab` and `backtab`. This may be a little controversial given how emacs tries to be super smart, but the hassle is often not worth it IMO.

And remember, this is merely a starting point for discussion. Everything can be discussed :)
Have a nice day!

purcell commented on elm-mode.el:
```diff
@@ -106,69 +231,47 @@ Find the roots of this function in the c-awk-mode."
 (define-derived-mode elm-mode prog-mode "Elm"
   "Major mode for editing Elm source code."
-  (setq-local indent-tabs-mode nil)
+  :group 'elm
```
This is implicit.

theothornhill replied:
Ok!

purcell commented on elm-mode.el:
```diff
@@ -106,69 +231,47 @@ Find the roots of this function in the c-awk-mode."
   "Major mode for editing Elm source code."
-  (setq-local indent-tabs-mode nil)
+  :group 'elm
+  :syntax-table elm--syntax-table
```
This would be implicit if the syntax table were to be called `elm--syntax-table`, which can safely be the case.

theothornhill replied:
Ok!

purcell reviewed (commented):
Thanks, some good stuff here, but I feel like it's starting in the wrong place and likely to disrupt many workflows. I don't think we should *start* by assuming LSP support when there is working support for direct integration of common tools like `elm-format` that might suit many people's workflows better for now. Also `imenu` support is generally straightforward and is reasonably expected of a major mode.

Simplifying/improving indentation code etc. is a less controversial first step, as would be dropping support for 0.18, including all the elm-oracle machinery, which won't work with 0.19+ anyway.

theothornhill commented:
Firstly - thanks for the review so far!

> 
> 
> Thanks, some good stuff here, but I feel like it's starting in the wrong place and likely to disrupt many workflows. I don't think we should _start_ by assuming LSP support when there is working support for direct integration of common tools like `elm-format` that might suit many people's workflows better for now. Also `imenu` support is generally straightforward and is reasonably expected of a major mode.

Right - thats fine by me. Does the `elm-package` integration work on your end? Has it ever worked? Right now I'm getting some `(wrong-type-argument listp (0ui/elm-task-parallel . ["1.0.0" "1.0.1" "1.0.2" "2.0.0"]))` when trying to invoke `M-x elm-package-catalog`. I'm wondering because if it isn't working, that whole thing could go, since probably not many people are using it 😄, otherwise, I could try to debug it a little. It seems like `let-alist` wants an `alist`, but is getting a normal cons pair, which isnt a list.



> 
> Simplifying/improving indentation code etc. is a less controversial first step, as would be dropping support for 0.18, including all the elm-oracle machinery, which won't work with 0.19+ anyway.

Ok sure. Will create a new commit when I get some free time on my hands. `elm-indent` is a beast, so not really sure how to approach that.

purcell commented:
> Right - thats fine by me. Does the `elm-package` integration work on your end? Has it ever worked? Right now I'm getting some `(wrong-type-argument listp (0ui/elm-task-parallel . ["1.0.0" "1.0.1" "1.0.2" "2.0.0"]))` when trying to invoke `M-x elm-package-catalog`. I'm wondering because if it isn't working, that whole thing could go, since probably not many people are using it 😄, otherwise, I could try to debug it a little. It seems like `let-alist` wants an `alist`, but is getting a normal cons pair, which isnt a list.

Hmm, I'd probably get the same result as you locally. Ideally we'd keep that functionality (but working, obviously) if possible, because it seems broadly useful.

> > Simplifying/improving indentation code etc. is a less controversial first step, as would be dropping support for 0.18, including all the elm-oracle machinery, which won't work with 0.19+ anyway.
> 
> Ok sure. Will create a new commit when I get some free time on my hands. `elm-indent` is a beast, so not really sure how to approach that.

Maybe make an `elm-indent-new` for now, since I think you were talking about making it pluggable initially?

theothornhill commented:
Sure, I'll make `elm-indent-simple`, and try to look into the `elm-package` stuff after that. 

Maybe we'll just add the `elm-indent-simple` as a part of this pr, in addition to removing <0.19 stuff, then start working on the other stuff incrementally?

purcell commented:
Yeah, sure, or separate small PRs are equally fine, and often easier to review and just merge in successively.

theothornhill commented:
Is the current PR more akin to what you want?

```elisp
(use-package elm-mode
  :ensure nil
  :config
  (setq elm-mode-enable-indent-simple t))
```

This snippet changes the indentation method, and it defaults to the current method.

It should be quite non-intrusive as for now, I think :)

purcell commented on elm-mode.el:
```diff
@@ -146,33 +154,42 @@ Find the roots of this function in the c-awk-mode."
+    (progn
+      (remove-hook 'elm-mode-hook 'elm-indent-simple-mode)
+      (add-hook 'elm-mode-hook 'elm-indent-mode)))
+  )
```
This stranded paren is why the build is failing at the `package-lint` setp btw.

purcell reviewed (commented):
Nice! I spammed a few review notes. :-)

theothornhill commented on elm-interactive.el:
```diff
@@ -925,230 +915,6 @@ EXPOSING"
 
 (autoload 'popup-make-item "popup")
```
Yeah, but i _think_ the other requires are needed.

theothornhill commented on elm-indent-simple.el:
```diff
@@ -0,0 +1,174 @@
+     (forward-line -1)
+     (end-of-line)
+     (forward-comment (- (point)))
+     (looking-back (regexp-opt ',tokens) nil)))
```
Not sure I completely understand what you mean here, but maybe the latest commit did want you want?

theothornhill commented:
I think I've addressed all your comments - thanks :)

Also, now you have to do this in your config:

```elisp

(use-package elm-mode
  :init
  (setq elm-mode-indent-mode 'elm-indent-simple-mode))

```
What do you think?

theothornhill commented:
@purcell Hey there, did you get a chance to take a look :)

purcell commented on elm-mode.el:
```diff
@@ -143,36 +146,46 @@ Find the roots of this function in the c-awk-mode."
 ;; We enable intelligent indenting, but users can remove this from the
 ;; hook if they prefer.
-(add-hook 'elm-mode-hook 'elm-indent-mode)
+(add-hook 'elm-mode-hook elm-mode-indent-mode)
```
Needs to be the quoted symbol here
```suggestion
(add-hook 'elm-mode-hook elm-mode-indent-mode)
```

purcell commented on elm-indent-simple.el:
```diff
@@ -0,0 +1,171 @@
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Indentation
```
```suggestion
(require 'cl-lib)

;;; Indentation
```

purcell commented on elm-indent-simple.el:
```diff
@@ -0,0 +1,171 @@
+       (forward-line -1)
+       (current-indentation))
+     (progn
+       (forward-line -1)
```
Note that sometimes there may not *be* one or even two previous lines, in which case `(forward-line -1)` would fail silently and you'd just get the indentation of the currently-reached line.

purcell commented on elm-indent-simple.el:
```diff
@@ -0,0 +1,171 @@
+
+(defvar elm-indent-simple-mode-map
+  (let ((map (make-sparse-keymap)))
+    (define-key map (kbd "C-c TAB") 'elm-indent-simple-indent-line)
```
I wonder if these are the best bindings: generally you don't want to override `indent-for-tab-command`, because completion can be hooked into it (see the `tab-always-indent` variable). I feel like `TAB` should therefore be left unchanged, `<backtab>` can still do the `backward` thing.

purcell reviewed (commented):
I'm comfortable merging this now and then we can tweak stuff incrementally. Need to follow up and update corresponding docs that might claim 0.18 compatibility.

theothornhill commented:
Great, thanks! I will update the docs:)

purcell commented:
I updated the README after this but there might be other occurrences.
